mipengine/controller/algorithm_executor.py

```diff
@@ -361,12 +231,230 @@ def get_run_udf_query(
+        message = f"Tables: {tables_schemas} do not have a common schema"
+        super().__init__(message)
+
+
+class InconsistentUDFResultSizeException(Exception):
+    def __init__(self, result_tables: Dict[int, List[Tuple["_Node", TableName]]]):
+        message = (
+            f"The following udf execution results on multiple nodes should have "
+            f"the same number of results.\nResults:{result_tables}"
+        )
+        super().__init__(message)
+
+
+class InconsistentShareTablesValueException(Exception):
+    def __init__(
+        self, share_list: Union[bool, List[bool]], result_tables: Union[list, dict]
+    ):
+        message = (
+            f"The size of the {share_list=} does not match the size of the "
+            f"{result_tables=}"
+        )
+        super().__init__(message)
+
+
+class AlgorithmExecutor:
+    def __init__(
+        self,
+        algorithm_execution_dto: AlgorithmExecutionDTO,
+        nodes_tasks_handlers_dto: NodesTasksHandlersDTO,
+    ):
+        self._logger = ctrl_logger.get_request_logger(
+            context_id=algorithm_execution_dto.context_id
+        )
+        self._nodes_tasks_handlers_dto = nodes_tasks_handlers_dto
+        self._algorithm_execution_dto = algorithm_execution_dto
+
+        self._context_id = algorithm_execution_dto.context_id
+        self._algorithm_name = algorithm_execution_dto.algorithm_name
+
+        self._global_node = None
+        self._local_nodes = []
+        self._execution_interface = None
+
+    def _instantiate_nodes(self):
+
+        # instantiate the GLOBAL Node object
+        self._global_node = _Node(
+            context_id=self._context_id,
+            node_tasks_handler=self._nodes_tasks_handlers_dto.global_node_tasks_handler,
+        )
+
+        # Parameters for the creation of the view tables in the db. Each of the LOCAL
+        # nodes will have access only to these view tables and not on the primary data
+        # tables
+        initial_view_tables_params = {
+            "commandId": get_next_command_id(),
+            "pathology": self._algorithm_execution_dto.algorithm_request_dto.inputdata.pathology,
+            "datasets": self._algorithm_execution_dto.algorithm_request_dto.inputdata.datasets,
+            "x": self._algorithm_execution_dto.algorithm_request_dto.inputdata.x,
+            "y": self._algorithm_execution_dto.algorithm_request_dto.inputdata.y,
+            "filters": self._algorithm_execution_dto.algorithm_request_dto.inputdata.filters,
+        }
+
+        # instantiate the LOCAL Node objects
+        self._local_nodes = [
+            _Node(
+                context_id=self._context_id,
+                node_tasks_handler=node_tasks_handler,
+                initial_view_tables_params=initial_view_tables_params,
+            )
+            for node_tasks_handler in self._nodes_tasks_handlers_dto.local_nodes_tasks_handlers
+        ]
+
+    def _instantiate_algorithm_execution_interface(self):
```

Comment:
Why not implement a `SingleNodeAlgorithmExecutionInterface`?

It would be much cleaner and easier to understand how it works, instead of using the `_is_single_node_execution` in multiple places to change the flow of the execution.

Most methods would be common, in a mother class, and then implement the differences, for example `_handle_table_sharing_locals_to_global` and `_handle_table_sharing_global_to_locals`. 

Even better, these methods don't need to exist at all in the `SingleNodeAlgorithmExecutionInterface` if `run_udf_on_global_node` and `run_udf_on_local_nodes` are written with single node execution in mind.

Reply:
done